src/Firestore/PathTrait.php

```diff
@@ -0,0 +1,197 @@
+    }
+
+    /**
+     * @param string $projectId The project ID.
+     * @param string $database The database name.
+     */
+    private function databaseName($projectId, $database)
+    {
+        return sprintf('projects/%s/databases/%s', $projectId, $database);
+    }
+
+    /**
+     * Get the database name from a path.
+     *
+     * @param string $name
+     * @return string
+     */
+    private function databaseFromName($name)
+    {
+        $parts = explode('/databases/', $name);
+        return $parts[0] . '/databases/' . explode('/', $parts[1])[0];
+    }
+
+    /**
+     * Determine whether the given path is a document.
+     *
+     * @param string $name
+     * @return bool
+     */
+    private function isDocument($name)
+    {
+        if (!$this->isRelative($name)) {
+            $name = $this->relativeName($name);
+        }
+
+        $parts = $this->splitName($name);
+        return count($parts) > 0 && count($parts) % 2 === 0;
+    }
+
+    /**
+     * Determine whether the given path is a collection.
+     *
+     * @param string $name
+     * @return bool
+     */
+    private function isCollection($name)
+    {
+        if (!$this->isRelative($name)) {
+            $name = $this->relativeName($name);
+        }
+
+        $parts = $this->splitName($name);
+        return count($parts) % 2 === 1;
+    }
+
+    /**
+     * Determine whether the given path is relative or absolute.
+     *
+     * @param string $name
+     * @return bool
+     */
+    private function isRelative($name)
+    {
+        $parts = $this->splitName($name);
+        return count($parts) > 0 && $parts[0] !== 'projects';
+    }
+
+    /**
+     * Split a path into pieces at the separator (`/`).
+     *
+     * @param string $name
+     * @return array
+     */
+    private function splitName($name)
+    {
+        return explode('/', trim($name, '/'));
+    }
+
+    /**
+     * Return the identifier from a path.
+     *
+     * @param string $name
+     * @return string|null
+     */
+    private function pathId($name)
+    {
+        $parts = $this->splitName($name);
+        if ($parts[0] === '') {
+            return null;
+        }
+
+        return end($parts);
+    }
+
+    /**
+     * Append a new element to a given path.
+     *
+     * @param string $name
+     * @param string $child
+     * @return string
+     */
+    private function childPath($name, $child)
+    {
+        return $name . '/' . $child;
+    }
+
+    /**
+     * Get the current path's parent.
+     * @param string $name
+     * @return string
+     */
+    private function parentPath($name)
+    {
+        $parts = $this->splitName($name);
+        array_pop($parts);
+
+        return implode('/', $parts);
+    }
+
+    /**
+     * Create a random name.
+     *
+     * @param string $name
+     * @return string
+     */
+    private function randomName($name)
+    {
+        $rand = preg_filter('/[^a-zA-Z0-9]{0,}/', '', base64_encode(random_bytes(30)));
```

Comment:
So I know that this is never going to be an issue, but since I needed to find something I could comment on in this line, I decided to point out that Base64 includes both `+` and `/`.